### PR TITLE
Nobodyliveshere-code-review

### PR DESCRIPTION
**Pull request by NobodyLivesHere**

### CODE REVIEW

File: habitat_code/nobodyliveshere-kmeans.py [hunk 1]
```diff
@@ -0,0 +1,804 @@
...
+import pandas as pd
+import pickle
+import rioxarray as rxr
+import xarray as xr
+
+
+from osgeo import gdal
+from shapely.geometry import Polygon
+from sklearn.cluster import KMeans
+
```
Comment (edchanonline) [inline habitat_code/nobodyliveshere-kmeans.py:106]: PEP8 Style - Probably something like this.

```
import hashlib
import os
import pickle

import earthaccess as eta
import earthpy as et
import geopandas as gpd
import geoviews.tile_sources as gt
import hvplot.pandas
import numpy as np
import pandas as pd
import rioxarray as rxr
import xarray as xr
from osgeo import gdal
from shapely.geometry import Polygon
from sklearn.cluster import KMeans
```

File: habitat_code/nobodyliveshere-kmeans.py [hunk 2]
```diff
@@ -0,0 +1,804 @@
...
+
+
+os.environ["GDAL_HTTP_MAX_RETRY"] = "5"
+os.environ["GDAL_HTTP_RETRY_DELAY"] = "1"
+
+
+# %%
+import sys
+
```
Comment (eggvoice) [inline habitat_code/nobodyliveshere-kmeans.py:115]: Move to top

File: habitat_code/nobodyliveshere-kmeans.py [hunk 3]
```diff
@@ -0,0 +1,804 @@
...
+from sklearn.cluster import KMeans
+
+
+
+os.environ["GDAL_HTTP_MAX_RETRY"] = "5"
+os.environ["GDAL_HTTP_RETRY_DELAY"] = "1"
+
+
+# %%
+import sys
```
Comment (edchanonline) [inline habitat_code/nobodyliveshere-kmeans.py:114]: Move to top

File: habitat_code/nobodyliveshere-kmeans.py [hunk 4]
```diff
@@ -0,0 +1,804 @@
...
+#   2. Compile all the information you collected into a GeoDataFrame
+
+# %%
+import geopandas as gpd
+from shapely.geometry import Polygon
+import re
+
+def compile_granule_info(search_results):
+    granules_info = []
+    desired_bands = ['01','02', '03', '04', '05', '06', '07', '09']  # Bands to include
```
Comment (edchanonline) [inline habitat_code/nobodyliveshere-kmeans.py:282]: Add comments to describe these bands. Alternatively, use a dictionary to map key:value pairs.

File: habitat_code/nobodyliveshere-kmeans.py [hunk 5]
```diff
@@ -0,0 +1,804 @@
...
+                if 'Fmask.tif' in url_info['URL']:
+                    fmask_url = url_info['URL']
+                elif url_info['URL'].endswith('.tif'):
+                    band_match = re.search(r'B(\d{2})\.tif', url_info['URL'])
+                    if band_match and band_match.group(1) in desired_bands:
+                        band_urls.append(url_info['URL'])
+
+        if not fmask_url:
+            print(f"No Fmask URL found for granule: {granule_id}")
+            continue  # Skip this granule if no Fmask URL is found
```
Comment (edchanonline) [inline habitat_code/nobodyliveshere-kmeans.py:307]: Interesting! Didn't know this is a possible scenario.

File: habitat_code/nobodyliveshere-kmeans.py [hunk 6]
```diff
@@ -0,0 +1,804 @@
...
+
+        for url in band_urls:
+            tile_id = re.search(r'T(\d{2}[A-Z]{3})', url).group(1) if re.search(r'T(\d{2}[A-Z]{3})', url) else None
+            band_number = int(re.search(r'B(\d{2})\.tif', url).group(1)) if re.search(r'B(\d{2})\.tif', url) else None
```
Comment (edchanonline) [inline habitat_code/nobodyliveshere-kmeans.py:311]: Nice regex statements 🔥

File: habitat_code/nobodyliveshere-kmeans.py [hunk 7]
```diff
@@ -0,0 +1,804 @@
...
+
+    # Check if the mask was applied correctly
+    print(f"Masked band data stats - Min: {test_masked_band_da.min().item()}, Max: {test_masked_band_da.max().item()}, Mean: {test_masked_band_da.mean().item()}, NaNs: {np.isnan(test_masked_band_da).sum().item()}")
+
+    return test_masked_band_da
+
+
+# %%
+# Assuming 'band_url' is the column name where the band paths are stored
+mask_bits = [1, 2, 3, 6, 7]
```
Comment (edchanonline) [inline habitat_code/nobodyliveshere-kmeans.py:383]: It'd be hard to guess what you are masking without context

File: habitat_code/nobodyliveshere-kmeans.py [hunk 8]
```diff
@@ -0,0 +1,804 @@
...
+scale_factor = 0.0001
+
+# Assuming granule_gdf is your GeoDataFrame with granule information
+# Add a new column for processed data if it doesn't exist
+if 'processed_data' not in granule_gdf.columns:
+    granule_gdf['processed_data'] = pd.Series(dtype=object)
+
+# Function to process a single band
+@cached(lambda granule_id, band_key, band_url, mask_bits, scale_factor, mask, watershed: f"{granule_id}-{band_key}")
```
Comment (edchanonline) [inline habitat_code/nobodyliveshere-kmeans.py:477]: Cool! 🔥

File: habitat_code/nobodyliveshere-kmeans.py (hunk 7 above)
Comment (advythr) [inline habitat_code/nobodyliveshere-kmeans.py:383]: It might be helpful for the reader to include here in the code more explicitly why the particular bits are being masked out.

File: habitat_code/nobodyliveshere-kmeans.py [hunk 9]
```diff
@@ -0,0 +1,804 @@
...
+#          ```
+# 
+#     2. For each band that starts with 'B':
+#         1. Open the band, crop, and apply the scale factor
+#         2. Name the DataArray after the band using the `.name` attribute
+#         3. Apply the cloud mask using the `.where()` method
+#         4. Store the DataArray in your data structure (e.g. adding a GeoDataFrame column with the DataArray in it. Note that you will need to remove the rows for unused bands)
+
+# %%
+import pandas as pd
```
Comment (advythr) [inline habitat_code/nobodyliveshere-kmeans.py:463]: To confirm to PEP-8 this should be moved to the top

File: habitat_code/nobodyliveshere-kmeans.py [hunk 10]
```diff
@@ -0,0 +1,804 @@
...
+            band_number = int(band_number)
+        if isinstance(band_number, int) and band_url.endswith('.tif') and 'B' in band_url:
+            band_key = 'B' + str(band_number)
+            processed_data = process_band(granule_id, band_key, band_url, mask_bits, scale_factor, mask, watershed)
+            granule_gdf.at[_, 'processed_data'] = processed_data
+            print(f"Processed data for {granule_id} - {band_key}: {granule_gdf.at[_, 'processed_data']}")
+
+        counter += 1  # Increment the counter after processing each row
+
+        if counter >= num_rows_to_process:
```
Comment (advythr) [inline habitat_code/nobodyliveshere-kmeans.py:536]: Interesting way of doing this, cool!

File: habitat_code/nobodyliveshere-kmeans.py [hunk 11]
```diff
@@ -0,0 +1,804 @@
...
+    .dropna()
+)
+
+# Assuming df is your DataFrame
+kmeans_model_df = kmeans_model_df.drop('spatial_ref', axis=1, level=0)
+
+
+
+# %%
+# Using an elbow plot to find optimum number K
```
Comment (advythr) [inline habitat_code/nobodyliveshere-kmeans.py:727]: I've never heard of this kind of plot, cool!

File: habitat_code/nobodyliveshere-kmeans.py [hunk 12]
```diff
@@ -0,0 +1,804 @@
...
+
+    plt.figure(figsize=(6, 6))
+    im = plt.imshow(data_normalized, cmap='gray')
+    plt.title(title)
+    plt.colorbar(im)
+    plt.show()
+
+# Assuming granule_gdf is your GeoDataFrame and it has been properly prepared
+# Extract processed data for bands B02, B03, and B04
+band_b02_data = granule_gdf.loc[granule_gdf['band_number'] == 2, 'processed_data'].iloc[0]
```
Comment (advythr) [inline habitat_code/nobodyliveshere-kmeans.py:569]: This could possibly be in a loop or function

**Review by advythr — COMMENTED** [review]
Overall this code is easy to follow and modular. I recommend being a bit careful with the imports to conform to PEP-8 and turning some things into loops for conciseness.

File: habitat_code/nobodyliveshere-kmeans.py (hunk 11 above)
Comment (NobodyLivesHere) [reply habitat_code/nobodyliveshere-kmeans.py:727]: Yeah, I found a youtube channel of a guy who does tutorials on using similar methods for petroleum exploration.  He showcased that method in a k-means clustering tutorial and I found it to work quite well.

File: habitat_code/nobodyliveshere-kmeans.py (hunk 4 above)
Comment (NobodyLivesHere) [reply habitat_code/nobodyliveshere-kmeans.py:282]: Yeah, I should have gone back and clarified what the code was doing at certain points.  I think in this scenario I was having issues with the downloads so I was tinkering with how many bands I could fit without blowing up the file size.

File: habitat_code/nobodyliveshere-kmeans.py (hunk 5 above)
Comment (NobodyLivesHere) [reply habitat_code/nobodyliveshere-kmeans.py:307]: It's unlikely. I mostly included it as an edge case for debugging in case the code did not fetch the fmask urls properly

File: habitat_code/nobodyliveshere-kmeans.py (hunk 6 above)
Comment (NobodyLivesHere) [reply habitat_code/nobodyliveshere-kmeans.py:311]: I've been trying to get the hang of using them more, since a lot of the objects we work with have different file naming conventions

File: habitat_code/nobodyliveshere-kmeans.py (hunk 7 above)
Comment (NobodyLivesHere) [reply habitat_code/nobodyliveshere-kmeans.py:383]: Think this was just for debugging

File: habitat_code/nobodyliveshere-kmeans.py (hunk 10 above)
Comment (NobodyLivesHere) [reply habitat_code/nobodyliveshere-kmeans.py:536]: I had set it up that way to test out different numbers of granules because of the amount of trouble I had processing the individual ones. That way I wouldn't have to create another object or break  a for loop.

**Review by NobodyLivesHere — COMMENTED** [review]
Seems mostly that I need to be more explicit in explaining the features being worked with and to look at shorter ways to achieve certain tasks by putting more code into functions to make it more readable.
